posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,160 @@
+
+Today, it takes around ~35 minutes for Linux and macOS tests to pass on GitHub Actions. With Scipy’s migration efforts towards the Meson build system, it takes around ~25 minutes for the CI to pass the build, install, and tests. Our primary focus was to reduce the overall time to achieve this. But finding workarounds for the same was particularly challenging.
+
+John initially proposed the idea of caching the source and the build directory together. We can restore the cache during the workflow and use remote sync to update the subset of changed files. It would allow us to perform an incremental build.
+
+While discussing with Ralf, I realized that using timestamp-based caching would be the way to go forward. My initial approach was to cache the build directory, which was nowhere effective. My conversation with Meson developers helped me realize that using a compiler caching tool like `ccache` or `sccache` would be the way forward to cache the build targets.
+
+To make sure that dependencies are being cached, I capitalized on the [cache action](https://github.com/actions/cache). GitHub Workflow runs often reuse the same downloaded dependencies from one run to another. GitHub’s cache actions allow us to retrieve cache identified by a unique key, for which I used a unique timestamp-based caching approach.
+
+While digging around the documentation, I discovered that `ccache` is natively supported by Meson and hence would be a better choice. I found that the build targets are exactly being cached by debugging around, and I can use the cache action to cache them. Finally, I made it work, thus effectively reducing the build time by nearly ~70% and reducing the CI time.
+
+In addition, I also worked on caching PyPi dependencies using the cache actions that allowed us to shave off some of the time required during the Python dependency installation. [PR #63](https://github.com/rgommers/scipy/pull/63) containing the commits for the code necessary to set up build caching as part of the GitHub Action workflow for Linux.
+
+## Implementing CI job for macOS
+
+With the GitHub Action workflows available for Linux, it seemed reasonable that we would like to test the same on macOS. This [issue](https://github.com/rgommers/scipy/issues/64) by [Rohit Goswami](https://github.com/HaoZeke), where he documents the problems with the macOS build, made sense that we can implement a workflow with little to almost no tweaks. However, I wanted to implement the conda-based approach rather than putting forward a native build.
+
+GitHub Actions allows us to perform macOS builds quite generously, and creating a workflow for the same was relatively easy. After the same, we had to make a couple of iterations to verify if the compiler caching exactly works and how performant it is compared to a fresh build. I also implemented conda caching for faster installs using [Mamba](https://github.com/mamba-org/mamba), which reduced the effective build time by 30%.
```

Comment:
Maybe just to avoid repetition:
```suggestion
GitHub Actions allows us to perform macOS builds quite generously, and creating a workflow for the same was relatively easy. After this, we had to make a couple of iterations to verify if the compiler caching exactly works and how performant it is compared to a fresh build. I also implemented conda caching for faster installs using [Mamba](https://github.com/mamba-org/mamba), which reduced the effective build time by 30%.
```